### ISSUE TITLE
avoid row break for long tables

### ISSUE THREAD

**Issue by nimish13** [wkhtmltopdf/wkhtmltopdf#3195]
I'm currently using wkhtmltopdf to generate large PDFs that includes a table which may be long enough to be shown on a single page. We were having issues getting rows to break properly at the end of each page.

I am following the below structure
```
html
<table>
   <tr>
     <td>
       <table>
          <tbody>
             <tr>
               <th></th>
             <tr>
             <tr>
                  <td></td>
             <tr>
                 .
                 .
                 .
                 .  
             <tr>
                 <td></td>
             <tr>
         </tbody>
         <tfoot>
             <tr>
             </tr>
                 .
                 .
                 .
             <tr>
             </tr>
               
         </tfoot>
       </table>
      </td>
  </tr>
</table>
```

We have tried using ```thead { display: table-header-group; }
tfoot { display: table-row-group; }
tr { page-break-inside: avoid; }``` and also other solutions provided on internet
But none is effective till now

Please help me in overcoming this issue.
Thanks in advance.

**Comment by PhilterPaper** [comment]
It would be helpful to tell us _how_ the table is (improperly) breaking, and how you would like it (or expect it) to behave. There have been many discussions here about how to keep a table from breaking in the middle of a row -- if that is the problem, you should look there first.

**Comment by nimish13** [comment]
Please find the attachment below.
[invoice.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/581356/invoice.pdf)

At the every page end the row is breaking and getting split on two pages

**Comment by PhilterPaper** [comment]
Ah yes, the problem of breaking in the middle of a _line_, as you show, has come up before. I don't recall there being any solution that worked for everyone. The problem is that the rendering engine is WebKit, and it's just not suitable for the purpose. It expects one continuous page of output, and doesn't work well when you need to split its output into pages.

Some users have had _some_ success with **page_break_inside: avoid** for rows, to split the table at a row, but that doesn't seem to always work.

**Comment by nimish13** [comment]
@PhilterPaper  Thanks for replying.
So there is no solution to this problem? Not even any hack?

**Comment by nimish13** [comment]
@ashkulz - Please provide a solution to this issue if any.

**Comment by spacemudd** [comment]
@PhilterPaper I spent right now 3 days scouring the internet for a solution until I saw yours. Holy crap. I almost gave up.
![fixed_](https://cloud.githubusercontent.com/assets/20286281/20227381/25d7ab50-a85d-11e6-81d6-fac6e095e75d.png)

**Comment by PhilterPaper** [comment]
@nimish13, sometimes people report success using CSS to prevent page breaks inside of rows, but you say you've already tried that. Are you using a reasonably up to date version of wkHTMLtoPDF? If you are, I have no idea of anything else you could try.

**Comment by nimish13** [comment]
@PhilterPaper Yes am using latest version of wkhtmltopdf i.e 12.3.3. Thanks for your time.

**Comment by alloylab** [comment]
@nimish13 Have you tried...
` tr {
	page-break-inside: avoid !important;
}` 

with an !important tag?

Also, what OS is wkhtmltopdf running on? Did you download it or compile it yourself? If you compiled it yourself are you using a patched QT version?

**Comment by susajsnair** [comment]
So far no fixes found. :(

**Comment by alloylab** [comment]
@susajsnair post in #2141

**Comment by ithan1985** [comment]
that can't be true

**Comment by ghoshsukanya** [comment]
[Tax Invoice.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2707112/Tax.Invoice.pdf)

My table is breaking in PDF. Please give a solution.
table tr { page-break-inside: avoid; } is not working.

**Comment by assyk** [comment]
tr { page-break-inside: avoid; page-break-after: auto; } 
+ Try with adding div inside the <td> tag and set it: display: inline-block. Only block and inline-block elements can be calculated correctly on page break

**Comment by jalpaks** [comment]
I have tried with all possibilities (more than 10) which mentioned above and another references , but unfortunately on server it won't work. Finally instead of table I have used div (for loop instead of "tr" use "<div class="table-row">") structure which is give table looks. and it working fine for me.
<head>
<style>
.page-break-table {
width: 100%;
display: block;
}
 .table-head {
 border-bottom: 1px solid #E5E5E5;
}
.table-row {
 width: 100%;
}
  
.table-row>* {
vertical-align: middle;
}
 .table-col {
 display: inline-block;
}
</style>
</head>
<body>
<div class="page-break-table">
    <div class="table-head">
        <div class="table-row">
            <div class="table-col" style="width:50%">your head content</div>
            <div class="table-col" style="width:50%">your head content</div>
        </div>
    </div>
    <div class="table-body">
        <div class="table-row">
            <div class="table-col" style="width:50%">your table body content</div>
            <div class="table-col" style="width:50%">your table body content</div>
        </div>
    </div>
</div>
</body>

**Comment by lenomatos** [comment]
I just remove table-responsive from table class and add

 tr {
      page-break-inside: avoid !important;
      page-break-after: auto !important;
  }

  @media print { .tr {
      page-break-inside: avoid !important;
      -webkit-column-break-inside: avoid;
      break-inside: avoid;
      -webkit-region-break-inside: avoid;
  }
  
  
  And it work right!
